### DISCUSSION TITLE
Retry with new authorization token

### DISCUSSION THREAD

**Discussion by paulius-miksys** [go-resty/resty#548] [Q&A] [answered]
I want to refresh my authorization token when it expires and I get 401. So I use `AddRetryCondition` (check if response is 401), `AddRetryHook` to update header and `SetRetryCount(1)`.

I set a new header with `SetHeader` in the func I pass to `AddRetryHook`. Interesting thing that resty still does a retry with an old authorization token even the header was successfully updated. As a result I get 401 again.

**Comment by kecci** [comment, marked as answer]
> I want to refresh my authorization token when it expires and I get 401. So I use `AddRetryCondition` (check if response is 401), `AddRetryHook` to update header and `SetRetryCount(1)`.
> 
> I set a new header with `SetHeader` in the func I pass to `AddRetryHook`. Interesting thing that resty still does a retry with an old authorization token even the header was successfully updated. As a result I get 401 again.

I'm trying to define your code like this:

```go
package main

import (
	"net/http"
	"time"

	"github.com/go-resty/resty/v2"
)

func main() {

	// Create a Resty Client
	client := resty.New()

	// Retries are configured per client
	client = client.
		// Set retry count to non zero to enable retries
		SetRetryCount(1).
		// You can override initial retry wait time.
		// Default is 100 milliseconds.
		SetRetryWaitTime(5 * time.Second).
		// MaxWaitTime can be overridden as well.
		// Default is 2 seconds.
		SetRetryMaxWaitTime(20 * time.Second).
		// SetRetryAfter sets callback to calculate wait time between retries.
		// Default (nil) implies exponential backoff with jitter
		AddRetryHook(
			func(r *resty.Response, err error) {
				client = client.SetHeader("Authorization", "auth-2")
			},
		)

	client = client.AddRetryCondition(
		// RetryConditionFunc type is for retry condition function
		// input: non-nil Response OR request execution error
		func(r *resty.Response, err error) bool {
			println(r.Request.Header.Get("Authorization"))
			println(r.StatusCode())
			return r.StatusCode() == http.StatusUnauthorized
		},
	)

	resp, _ := client.R().SetHeader("Authorization", "auth-1").Get("https://webhook.site/9e0d8a64-0ed1-4e28-ae32-ddd26fae5e1d")

	println(resp.Request.Header.Get("Authorization"))
}
```

Result:

```
auth-1
401
auth-1
401
auth-1
```

As you can see, using `SetHeader` in the `AddRetryHook` is not solve your case. Because it's not update the latest header.

### Resolution
In this case, you are want to update your header if got 401, right ?

You can add do-while function with these step:
1. While Request with old header.
2. Do Validate if 401.
3. While Request again with new one.
4. Do if 200 passed.

I hope this could be help your questions

**Comment by wasserholz** [comment]
The answer is unsatisfying because it requires workarounds to handle exponential back off etc.
It would be easier if there would be retry hook that could be configured per request

**Reply by danhawkins** [reply]
I found a way of doing this, by adding OnBeforeReqyest

```
	c.rac = c.rac.
		SetRetryCount(1).
		AddRetryCondition(func(r *resty.Response, err error) bool {
			return err != nil || r.StatusCode() == http.StatusUnauthorized
		}).
		AddRetryHook(func(r *resty.Response, err error) {
			ctx := context.Background()
			newToken, err := c.getAdminToken(ctx)
			if err != nil {
				rlog.Error("error getting new token", "error", err)
				return
			}
			c.adminTokenCache.Set(ctx, newToken)
		})

	c.rac.OnBeforeRequest(func(_ *resty.Client, req *resty.Request) error {
		token, err := c.getCachedAdminToken(req.Context())
		if err != nil {
			rlog.Error("error getting admin token", "error", err)
			return err
		}

		req.SetHeader("Authorization", "Bearer "+token.AccessToken)
		return nil
	})
	```
